pycode/test/test_epidata_modifyDataframeSeries.py

```diff
@@ -18,72 +18,95 @@
-        'test_col1': [11.0, 11.0, 3.0, 3.0, 6.0],
-        'test_col2': ['a', 'a', 'x', 'x', 't']})
+        {
+         'Date':
+         ['2021-01-06', '2021-01-06', '2021-01-06', '2021-01-07', '2021-01-07',
+          '2021-01-07', '2021-01-08', '2021-01-08', '2021-01-08', '2021-01-09',
+          '2021-01-09', '2021-01-09', '2021-01-10', '2021-01-10',
+          '2021-01-10'],
+         'test_col1': [12, 3, 6, 0, 3, 1, 4, 7, 11, 15, 19, 19, 27, 13, 5],
+         'test_col2': ['a', 'x', 't', 'a', 'b', 'a', 'x', 't', 'a', 'b', 'a', 'x', 't', 'a', 'b'],
+         'test_col3': [1, 0, 1, 9, 4, 3, 2, 1, 1, 1, 0, 6, 5, 3, 1],
+         'ID': [1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3]})
+    test_df2 = pd.DataFrame(
+        {
+         'Date':
+         ['2021-01-06', '2021-01-06', '2021-01-06', '2021-01-07', '2021-01-07',
+          '2021-01-07', '2021-01-08', '2021-01-08', '2021-01-08', '2021-01-09',
+          '2021-01-09', '2021-01-09', '2021-01-13', '2021-01-13',
+          '2021-01-13'],
+         'test_col1': [12, 3, 6, 0, 3, 1, 4, 7, 11, 15, 19, 19, 27, 13, 5],
+         'test_col2': ['a', 'x', 't', 'a', 'b', 'a', 'x', 't', 'a', 'b', 'a', 'x', 't', 'a', 'b'],
+         'test_col3': [1, 0, 1, 9, 4, 3, 2, 1, 1, 1, 0, 6, 5, 3, 1],
+         'ID': [1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3]})
+
 
     def setUp(self):
         self.setUpPyfakefs()
 
     def test_impute_and_reduce_df(self):
-        group_by_cols = {}
-        mod_cols = ''
-        # test fill date
-        test_df_new = mDfS.impute_and_reduce_df(
-            self.filldates_df, group_by_cols, mod_cols, impute='forward',
-            moving_average=0, min_date='2021-01-05', max_date='2021-01-08',
+
+        group_by_cols = {'ID': sorted(set(self.test_df1['ID'].unique()))}
+        mod_cols = ['test_col1', 'test_col3']
+
+        # test impute forward and fill dates with moving average = 3
+        df = mDfS.impute_and_reduce_df(
+            self.test_df1, group_by_cols, mod_cols, impute='forward',
+            moving_average=3, min_date='2021-01-05', max_date='2021-01-11',
             start_w_firstval=False)
-        pd_testing.assert_frame_equal(test_df_new, self.test_df1)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-06") & (df['ID'] == 3.0)]['test_col3'].item(), 1 + 1 / 3)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-06") & (df['ID'] == 3.0)]['test_col1'].item(), 2 + 1 / 3)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-05") & (df['ID'] == 2.0)]['test_col3'].item(), 0)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-05") & (df['ID'] == 2.0)]['test_col1'].item(), 0)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-11") & (df['ID'] == 3.0)]['test_col3'].item(), 1)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-10") & (df['ID'] == 1.0)]['test_col1'].item(), 23)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-10") & (df['ID'] == 1.0)]['test_col3'].item(), 3+2/3)
```

Comment:
```suggestion
        # test that values at new start days are zero since start_w_firstval=False
        self.assertAlmostEqual(df[(df['Date'] == "2021-01-05") & (df['ID'] == 2.0)]['test_col1'].item(), 0)
        self.assertAlmostEqual(df[(df['Date'] == "2021-01-05") & (df['ID'] == 2.0)]['test_col3'].item(), 0)

        # test that the values at first original date are obtained by the value itself plus the value right of it divided by 3
        # (6 + 1) / 3 = 2.33333333333
        self.assertAlmostEqual(df[(df['Date'] == "2021-01-06") & (df['ID'] == 3.0)]['test_col1'].item(), 2 + 1/3)
        # (3 + 1) / 3 = 1.333333333
        self.assertAlmostEqual(df[(df['Date'] == "2021-01-06") & (df['ID'] == 3.0)]['test_col3'].item(), 2 + 1 / 3)
        
        # check that last entry of original frame is computed by the value left of it and twice its value since impute='forward'
        # (15 + 27 + 27) / 3 = 23
        self.assertAlmostEqual(df[(df['Date'] == "2021-01-10") & (df['ID'] == 1.0)]['test_col1'].item(), 23)
        # (1 + 5 + 5) / 3 = 3 + 2/3
        self.assertAlmostEqual(df[(df['Date'] == "2021-01-10") & (df['ID'] == 1.0)]['test_col3'].item(), 3 + 2/ 3)
        
        # check that new date gets imputed the value the column had the day before because impute='forward'
        self.assertAlmostEqual(df[(df['Date'] == "2021-01-11") & (df['ID'] == 3.0)]['test_col3'].item(), 1)
```
It would be good to have this structured with comments